### DISCUSSION TITLE
Corner detection for newb

### DISCUSSION THREAD

**Discussion by quasiTriestino** [ImageMagick/ImageMagick#3995] [Help]
I have a stack of roughly 200 complex grayscale images, each on the order of 8000 x 8000 px. Some of these images have been transformed and placed on a black background. I'd like to get the pixel coordinates from each of the corners of each transformed image in order to back calculate what transformation was applied to the original raw image. I've been playing around with canny edges and hough-lines, but to no avail. Is there a simpler method using ImageMagick that returns corners of images such as these or could someone point my in the right direction?  

![example](https://user-images.githubusercontent.com/8270782/127497291-6450386b-90b1-40d1-b24a-fcbc3095bed3.JPG)

**Comment by snibgo** [comment]
What version of IM, on what platform?

Please post an actual "transformed" version, or paste a link to the file, or a zip of the file.

My initial thinking is to turn non-black into white, then use morphology. There are many possible methods, but simplifying the image is a good first step.

**Reply by quasiTriestino** [reply]
Sorry about that - I should have thought to mention it in my post: IM 7.1.0-4 on Windows. I'll link to a [compressed example image here](https://michaelchirillo.com/img/trial-image.zip) (the uncompressed file is about 60 mb). I was (I think?) able to simplify the image to white on the original black background by running `convert trial-image.tif -fill white +opaque black output.tif`, though I'm not sure if this is the correct direction I should be taking. Any advice would be very much appreciated!

**Comment by snibgo** [comment]
Yes, that transformation works well:
```
magick trial-image.tif -fill White +opaque Black x.tiff
```
The result has some noise, but it is good enough.

My webpage [Find four corners](http://im.snibgo.com/find4corn.htm) shows various methods for finding the corners in x.tiff. That page is now quite old, and better (faster) methods are now available.

For example:
```
call %PICTBAT%find4cornSub x.tiff my4corners x2.tiff
```
This takes 34 seconds on my computer with your image. It creates environment variables with the x,y coordinates of the corners at botom-left, top-right etc.
```
my4corners.BL.x=378
my4corners.BL.y=7611
my4corners.BR.x=7838
my4corners.BR.y=7626
my4corners.TL.x=235
my4corners.TL.y=158
my4corners.TR.x=7695
my4corners.TR.y=172
```
That script calls other scripts. See [Zipped BAT files](http://im.snibgo.com/zipbats.htm) for instructions.

**Reply by quasiTriestino** [reply]
This looks super promising! Now I'm in a fight to the death with these environment variables. Keep running into things like `'echoOffSave' is not recognized as an internal or external command, operable program or batch file` and I can't figure out what's the matter. I've set the %Util% to the folder according to this in the imsnibgoBats.txt file `%Util%\echoOffSave.bat` but no dice as of yet....

**Reply by quasiTriestino** [reply]
> The zip file also contains imsnibgoBats.txt, which lists the BAT files, and also shows in which directory they are installed. Most BAT files are installed in %PICTBAT%. If that directory is in your system path, you don't need to set that environment variable. Some BAT files are in %Util%, and BAT files expect that directory to be in the system path, so those scripts are called without a %Util% prefix. 

Ok, ok, still working on this.....

**Comment by snibgo** [comment]
My image processing scripts expect that other image processing scripts are in a directory pointed to by %PICTBAT%, and that general-purpose scripts are in a directory that is on the system path, so they can be called without giving the explicit path.

So "call echoOffSave" expects a file called "echoOffSave.bat" to be in a directory that is on the system path.

But "call %PICTBAT%setInOut" expects a file called "setInOut.bat" to be in %PICTBAT%, so PICTBAT should be an environment variable that contains an explicit path such as "C:\snibgostuff\".

**Reply by quasiTriestino** [reply]
So the last piece of this puzzle - I think/hope - is the %IM% variable which currently points to "C:\Progra~1\ImageMagick-7.1.0-Q16-HDRI\". where magick.exe is located. However, now things like `'C:\Progra~1\ImageMagick-7.1.0-Q16-HDRI\convert' is not recognized as an internal or external command, operable program or batch file.` are popping up. Thank you for being patient about this - I think it's pretty clear this isn't something I deal with on a daily basis.

**Comment by snibgo** [comment]
I wrote those scripts years ago, for IM v6, when the programs were "convert" and "identify". You mentioned that you ran "convert trial-image.tif -fill white +opaque black output.tif", so you seem to have those v6 programs.

Where is your convert.exe? Set IM to that directory.

**Reply by quasiTriestino** [reply]
I run "magick convert..." or "magick identify..." in command prompt, so I would have run "magick convert trial-image.tif -fill white +opaque black output.tif"

**Comment by snibgo** [comment]
A faster method, but one that requires re-building IM with my [Process modules](http://im.snibgo.com/customim.htm) is:
```
magick ^
  trial-image.tif ^
  -fill White +opaque Black ^
  -write mpr:INP +delete ^
  -size 2x2 xc:Black -fill White -draw "point 1,1" ^
  -write mpr:SUB +delete ^
  ( mpr:INP ^
    mpr:SUB ^
    -process 'srchimg noCrop verbose' ^
  ) ^
  ( mpr:INP ^
    ( mpr:SUB -rotate 90 ) ^
    -process 'srchimg noCrop verbose' ^
  ) ^
  ( mpr:INP ^
    ( mpr:SUB -rotate 180 ) ^
    -process 'srchimg noCrop verbose' ^
  ) ^
  ( mpr:INP ^
    ( mpr:SUB -rotate 270 ) ^
    -process 'srchimg noCrop verbose' ^
  ) ^
  NULL:
```
This takes 11 seconds, of which 7 seconds is searching.

The sub-images are 2x2, so a conventional multi-scale approach can't be used. However, for this case we could multi-scale (subsample) the main image and use the unmodified subimage at each scale. This might reduce the search time dramatically.

Another possible approach is to use the recent Accelerated Subimage-search NCC Compare (https://github.com/ImageMagick/ImageMagick/discussions/3950) which needs fftw. I haven't had a chance to test this yet.

**Comment by quasiTriestino** [comment]
Ok, I think we're getting somewhere now! I reinstalled IM with legacy tools (keeps convert.exe, which is now in my current %IM% variable). My working directory is imsnibgoBats which is sitting my desktop. (%PICTBAT% points to that directory.) I'm calling `call %PICTBAT%find4cornSub "c:/users/mchirillo/desktop/output.tif" my4corners x2.tiff`. Output.tif is also sitting on my desktop and is the result of running `magick convert trial-image.tif -fill white +opaque black output.tif` on the original image.

Still getting some hiccups, but fewer and fewer, which - I gotta tell you - is encouraging on a Friday evening. I'm now running up against this: `'C:\Users\mchirillo\Desktop\imsnibgoBats\subSrchFactRep' is not recognized as an internal or external command, operable program or batch file.`

I can see in find4cornSub.bat the subSrchFactRep is called with the %PICTBAT% env variable: `call %PICTBAT%subSrchFactRep %QUARTERS%[0] %CORNER% %SRCH_FACT% F4Csub`

Looking in imsnibgoBats for SubSrchFactRep and not seeing it, nor is it listed in the accompanying .txt file - I'm going to try to hunt for it now on your website.

**Reply by snibgo** [reply]
Oops, sorry, subSrchFactRep.bat and subSrchFact.bat are not in the zip file. I will correct that. Meanwhile, here they are:

subSrchFactRep.bat is:
```
rem Repeatedly calls subSrchFact until result is good, or factor is 1.

@if "%2"=="" findstr /B "rem @rem" %~f0 & exit /B 1

@setlocal enabledelayedexpansion

@call echoOffSave

call %PICTBAT%setInOut %1 ssfr

set SML_FILE=%2

set FACT=%3
if "%FACT%"=="." set FACT=
if "%FACT%"=="" set FACT=10

set ENV_PREF=%4
if "%ENV_PREF%"=="." set ENV_PREF=
if "%ENV_PREF%"=="" set ENV_PREF=ssfrENV

set THRESH_BAD=%5
if "%THRESH_BAD%"=="." set THRESH_BAD=
if "%THRESH_BAD%"=="" set THRESH_BAD=0.1

:loop
set %ENV_PREF%.BAD=

call %PICTBAT%subSrchFact %INFILE% %SML_FILE% %FACT% %ENV_PREF% %THRESH_BAD%

if "!%ENV_PREF%.BAD!"=="" goto finished

if %FACT%==1 goto finished

for /F "usebackq" %%L in (`%IM%identify ^
  -format "FACT=%%[fx:%FACT%<=2?1:%FACT%/2.0]" ^
  xc:`) do set %%L

echo %0: FACT=%FACT% bad=!%ENV_PREF%.BAD!

goto loop


:finished

echo %0: FACT=%FACT% bad=!%ENV_PREF%.BAD!

set ENV_LIST=%BASENAME%_ssfrEnv.lis
set %ENV_PREF% >%ENV_LIST%

call echoRestore

@endlocal & for /F %%L in (%ENV_LIST%) do @set %%L
```

subSrchFact.bat is:
```
rem Given large image %1 and very small image %2,
rem searches for small in large, via a resize of the large image, integer factor %3 (eg 2, 20).
rem Writes to environment prefix %4:
rem   .x .y .score .bad
rem %5 is threshold RMSE for badness. Default 0.1


@if "%2"=="" findstr /B "rem @rem" %~f0 & exit /B 1

@setlocal enabledelayedexpansion

@call echoOffSave

call %PICTBAT%setInOut %1 ssf

set ENV_PREF=%4
if "%ENV_PREF%"=="." set ENV_PREF=
if "%ENV_PREF%"=="" set ENV_PREF=ssfENV

set SML_FILE=%2

set FACT=%3
if "%FACT%"=="." set FACT=
if "%FACT%"=="" set FACT=10

set THRESH_BAD=%5
if "%THRESH_BAD%"=="." set THRESH_BAD=
if "%THRESH_BAD%"=="" set THRESH_BAD=0.1

set TMP_LRG=%BASENAME%_ssf.miff
set TMP_LRG2=%BASENAME%_ssf2.miff

for /F "usebackq" %%L in (`%IM%identify ^
  -format "FACT_PC=%%[fx:100/%FACT%]" ^
  xc:`) do set %%L

%IM%convert ^
  %INFILE% ^
  +repage ^
  -resize %FACT_PC%%% ^
  %TMP_LRG%


goto skip
%IM%compare ^
  -metric RMSE ^
  %TMP_LRG% ^
  %SML_FILE% ^
  -similarity-threshold 0 ^
  -dissimilarity-threshold 1 ^
  -subimage-search ^
  NULL:
:skip

for /F "usebackq tokens=1-4 delims=()@, " %%A in (`%IM%compare ^
  -metric RMSE ^
  %TMP_LRG% ^
  %SML_FILE% ^
  -similarity-threshold 0 ^
  -dissimilarity-threshold 1 ^
  -subimage-search ^
  NULL: 2^>^&1`) do (
  set score=%%B
  set fndX=%%C
  set fndY=%%D
)
echo %0: fndX=%fndX% fndY=%fndY% score=%score%

set CALC_SUB=^
crpX=%%[fx:int((%fndX%-2)*%FACT%+0.5)]\n^
crpY=%%[fx:int((%fndY%-2)*%FACT%+0.5)]\n^
crpH=%%[fx:int(5*%FACT%+w+1+0.5)]\n^
crpW=%%[fx:int(5*%FACT%+h+1+0.5)]

for /F "usebackq" %%L in (`%IM%identify ^
  -format "%CALC_SUB%" ^
  %SML_FILE%`) do set %%L

set BAD=
if %crpX% LSS 0 set BAD=1
if %crpY% LSS 0 set BAD=1

echo %0: CROP=-crop %crpW%x%crpH%+%crpX%+%crpY% +repage

%IM%convert ^
  %INFILE% ^
  +repage ^
  -crop %crpW%x%crpH%+%crpX%+%crpY% +repage ^
  %TMP_LRG2%

for /F "usebackq tokens=1-4 delims=()@, " %%A in (`%IM%compare ^
  -metric RMSE ^
  %TMP_LRG2% ^
  %SML_FILE% ^
  -similarity-threshold 0 ^
  -dissimilarity-threshold 1 ^
  -subimage-search ^
  NULL: 2^>^&1`) do (
  set fndScore=%%B
  set /A fndX=%%C+%crpX%
  set /A fndY=%%D+%crpY%
)
echo %0: fndX=%fndX% fndY=%fndY% fndScore=%fndScore%

if not "%BAD%"=="1" for /F "usebackq" %%L in (`%IM%identify ^
  -format "BAD=%%[fx:%fndScore%>%THRESH_BAD%?1:0]" ^
  xc:`) do (
  set %%L
  if "!BAD!"=="0" set BAD=
)



call echoRestore

endlocal & set %ENV_PREF%.x=%fndX%& set %ENV_PREF%.y=%fndY%& set %ENV_PREF%.score=fndScore& set %ENV_PREF%.BAD=%BAD%
```

**Comment by quasiTriestino** [comment]
!!!SUCCESS!!! 

`
my4corners.BL.x=378
my4corners.BL.y=7611
my4corners.BR.x=7838
my4corners.BR.y=7626
my4corners.TL.x=235   
my4corners.TL.y=158
my4corners.TR.x=7695
my4corners.TR.y=172
`

I cannot thank you enough for walking me through this - I am very, very appreciative.

**Comment by snibgo** [comment]
Good stuff. I have updated the zip file to contain those bat files, and added some new pages, and updated a load of others.

**Reply by quasiTriestino** [reply]
Great! I'm glad this has been productive for both ends of this conversation :)

**Comment by quasiTriestino** [comment]
One note: This setup gets stuck if the corner happens to be at the edge of the image. I've modified things a little bit to add a 100px black border about the entire image, so that this doesn't happen. Seems to be working fine for the time being.

**Reply by snibgo** [reply]
Thanks for the note. I'll put an explanation on the web page.
